### PR TITLE
Fix deadlock introduce by #538

### PR DESCRIPTION
**Pull request by sbernard31**

### CODE REVIEW

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java [hunk 1]
```diff
@@ -1041,27 +1041,32 @@ private Block2BlockwiseStatus getBlock2Status(final KeyUri key) {
...
 		synchronized (block1Transfers) {
-			Block1BlockwiseStatus removedTracker = block1Transfers.remove(key);
-			if (removedTracker != null) {
-				LOGGER.debug("removing block1 tracker [{}], block1 transfers still in progress: {}",
-						new Object[] { key, block1Transfers.size() });
-				removedTracker.setComplete(true);
-			}
-			return removedTracker;
+			removedTracker = block1Transfers.remove(key);
+			size = block1Transfers.size();
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java]: I hesitated to enclose this line with a if LOGGER.isDebugEnabled() but the size() is just an access to an attribute so I'm no sure if this is worth.

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java [hunk 2]
```diff
@@ -1041,27 +1041,32 @@ private Block2BlockwiseStatus getBlock2Status(final KeyUri key) {
...
-						new Object[] { key, block2Transfers.size() });
-				removedTracker.setComplete(true);
-			}
-			return removedTracker;
+			removedTracker = block2Transfers.remove(key);
+			size = block2Transfers.size();
+		}
+		if (removedTracker != null) {
+			LOGGER.debug("removing block2 tracker [{}], block2 transfers still in progress: {}",
+					new Object[] { key, size });
```
Comment (boaks) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java]: Can we remove the `new Object[]`? It seems not to be any longer required.

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java (hunk 1 above)
Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java]: Don't make it too complex. So leave it, as it is here.

**Comment by boaks** [comment]
Except of the new Object[], LGTM

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java (hunk 2 above)
Comment (sbernard31) [reply californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java]: I will remove the `Object[]`.

By the way do you understand the log message ?
First part ok : _removing block2 tracker [{}]_
Second part : _block2 transfers still in progress: {size}_ sounds a bit strange to me ...

Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java]: replace "still" with "left"?

**Pull request merged**